signature/plugin.go

```diff
@@ -221,50 +190,25 @@ func (s *pluginSigner) generateSignatureEnvelope(ctx context.Context, desc notat
 		)
 	}
 
-	// Check signatureEnvelope contains a valid JWSEnvelope.
-	var envelope notation.JWSEnvelope
-	if err = json.Unmarshal(resp.SignatureEnvelope, &envelope); err != nil ||
-		len(envelope.Payload) == 0 ||
-		len(envelope.Protected) == 0 ||
-		len(envelope.Signature) == 0 ||
-		len(envelope.Header.CertChain) == 0 {
-
-		return nil, errors.New("envelope content does not match envelope format")
-	}
-
-	// Check algorithm is supported.
-	var protected notation.JWSProtectedHeader
-	if err = decodeBase64URLJSON(envelope.Protected, &protected); err != nil {
-		return nil, fmt.Errorf("envelope protected header can't be decoded: %w", err)
-	}
-	if notation.NewSignatureAlgorithmJWS(protected.Algorithm) == "" {
-		return nil, fmt.Errorf("signing algorithm %q not supported", protected.Algorithm)
-	}
-
-	// Check descriptor subject is honored.
-	var payload notation.JWSPayload
-	err = decodeBase64URLJSON(envelope.Payload, &payload)
+	sigEnv, err := signer.NewSignatureEnvelopeFromBytes(resp.SignatureEnvelope, signer.MediaTypeJWSJson)
 	if err != nil {
-		return nil, fmt.Errorf("envelope payload can't be decoded: %w", err)
-	}
-	if !descriptorPartialEqual(desc, payload.Subject) {
-		return nil, errors.New("descriptor subject has changed")
+		return nil, err
 	}
 
-	// Check signatureEnvelope can be verified against signing certificate.
-	certs, err := parseCertChain(envelope.Header.CertChain)
+	sigInfo, err := sigEnv.Verify()
 	if err != nil {
 		return nil, err
 	}
-	err = verifyJWT(protected.Algorithm, envelope.Protected+"."+envelope.Payload, envelope.Signature, certs[0])
-	if err != nil {
-		return nil, err
+
+	var signedPayload notation.Payload
+	if err = json.Unmarshal(sigInfo.Payload, &signedPayload); err != nil {
+		return nil, fmt.Errorf("signed envelope payload can't be unmarshaled: %w", err)
 	}
 
-	// Check the the certificate chain conforms to the spec.
-	if err := verifyCertExtKeyUsage(certs[0], x509.ExtKeyUsageCodeSigning); err != nil {
-		return nil, fmt.Errorf("signing certificate does not meet the minimum requirements: %w", err)
+	if !descriptorPartialEqual(desc, signedPayload.TargetArtifact) {
+		return nil, errors.New("descriptor subject has changed")
```

Comment:
Envelope generator should not add any additional top level payload attributes, we should verify that.

Reply:
Added TODO.